File: 2025-09/2025-09-04/vowel_repeater.py
import re

def repeat_vowels(s):
    regex = re.compile(r"[aeiouAEIOU]")
    vowel_repeater_str = ''
    vowel_counter = 0;

    for char in s:
        vowel_repeater_element = char
        if re.match(regex, char):
            vowel_repeater_element += char.lower() * vowel_counter
            vowel_counter += 1
        vowel_repeater_str += vowel_repeater_element
        
    return vowel_repeater_str

print(repeat_vowels("hello world")) 
print(repeat_vowels("freeCodeCamp"))
print(repeat_vowels("AEIOU"))
print(repeat_vowels("I like eating ice cream in Iceland"))

# --- TEST SUITE ---

tests_text = '''
1. repeat_vowels("hello world") should return "helloo wooorld".
2. repeat_vowels("freeCodeCamp") should return "freeeCooodeeeeCaaaaamp".
3. repeat_vowels("AEIOU") should return "AEeIiiOoooUuuuu".
4. repeat_vowels("I like eating ice cream in Iceland") should return "I liikeee eeeeaaaaatiiiiiing iiiiiiiceeeeeeee creeeeeeeeeaaaaaaaaaam iiiiiiiiiiin Iiiiiiiiiiiiceeeeeeeeeeeeelaaaaaaaaaaaaaand".
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+?)\.?$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("--------------------------")
    print("🧪 Starting Verification...")
    print("--------------------------")

    all_passed = True
    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            all_passed = False
            fail_count += 1

    print("----------------------------");

    if (all_passed):
        print("🎉 SUCCESS: All tests PASSED.");
    else:
        print(f"⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED.");

run_tests(test_data)

File: 09 September/09-04/vowelRepeater.js
function repeatVowels(str) {
  const regex = /[aeiouAEIOU]/;
  let vowelRepeaterArr = str.split("");
  let vowelCounter = 0;

  vowelRepeaterArr = vowelRepeaterArr.map(element => {
    let vowelRepeaterElement = element;
    
    if (regex.test(element)) {
      for (let i = 0; i < vowelCounter; i ++) {
        vowelRepeaterElement += element.toLowerCase();
      }
    vowelCounter ++;
    }

    return vowelRepeaterElement;
  })

  return vowelRepeaterArr.join('');
}

console.log(repeatVowels("hello world")); 
console.log(repeatVowels("freeCodeCamp"));
console.log(repeatVowels("AEIOU"));
console.log(repeatVowels("I like eating ice cream in Iceland"));

// --- TEST SUITE ---

const testsText = `
1. repeatVowels("hello world") should return "helloo wooorld".
2. repeatVowels("freeCodeCamp") should return "freeeCooodeeeeCaaaaamp".
3. repeatVowels("AEIOU") should return "AEeIiiOoooUuuuu".
4. repeatVowels("I like eating ice cream in Iceland") should return "I liikeee eeeeaaaaatiiiiiing iiiiiiiceeeeeeee creeeeeeeeeaaaaaaaaaam iiiiiiiiiiin Iiiiiiiiiiiiceeeeeeeeeeeeelaaaaaaaaaaaaaand".
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+?)\.?$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("--------------------------");
  console.log("🧪Starting Verification...");
  console.log("--------------------------");

  let allPassed = true;
  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(test.output);
    let comparison;

    if (Array.isArray(testOutput))
      comparison = arraysEqual(functionCallOutput, testOutput);
    else
      comparison = functionCallOutput === testOutput;

    if (comparison) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    }
    else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${testOutput}\nGot: ${functionCallOutput}`);
      allPassed = false;
      failCount ++;
    }
  })

  console.log("----------------------------");

  if (allPassed)
    console.log("🎉SUCCESS: All tests PASSED.");
  else
    console.log(`⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`);
}

function arraysEqual(a, b) {
  if (a.length !== b.length)
    return false;

  for (const index in a) {
    if (Array.isArray(a[index])) {
      if (Array.isArray(b[index])) {
        if (!arraysEqual(a[index], b[index]))
          return false;
      }
      else 
        return false;
    }

    else if (a[index] !== b[index])
      return false;
  }

  return true;
}

runTests(testData);